Wikipedia/Code/Panels.swift

```diff
@@ -63,13 +63,28 @@ class ReLoginFailedPanelViewController : ScrollableEducationPanelViewController
 }
 
 class LoginOrCreateAccountToSyncSavedArticlesToReadingListPanelViewController : ScrollableEducationPanelViewController {
+    private var tappedPrimaryButton: Bool = false
+
     override func viewDidLoad() {
         super.viewDidLoad()
         image = UIImage.init(named: "reading-list-user")
         heading = WMFLocalizedString("reading-list-login-or-create-account-title", value:"Log in to sync saved articles", comment:"Title for syncing saved articles.")
         subheading = CommonStrings.readingListLoginSubtitle
         primaryButtonTitle = WMFLocalizedString("reading-list-login-or-create-account-button-title", value:"Log in or create account", comment:"Title for button to login or create account to sync saved articles and reading lists.")
     }
+    
+    override func primaryButtonTapped(_ sender: Any) {
+        tappedPrimaryButton = true
+        super.primaryButtonTapped(sender)
+    }
+    
+    override func viewWillDisappear(_ animated: Bool) {
+        guard !tappedPrimaryButton else {
```

Comment:
Can you say why you are avoiding calling super in this case? This will also prevent all other super classes from running cleanup code (like UIViewController).

Reply:
@nambatee this is probably the only thing I think would need to be at least commented on before merging. The other things are nice cleanup/questions to consider but non-blocking.

Reply:
yeah, I missed that, thanks! I wanted to avoid calling `dismissHandler` because if user dismisses the panel, I want the sync switch to be turned off but if they tap the button to 'log in', I don't want them to see the switch being turned off right before they go to the login view. With the superclass calling `dismissHandler` in `viewWillDisappear`, I didn't have a good way to prevent that. I decided to have a flag in my own panel, override `viewWillDisappear` and decide whether I want to let `super` call `dismissHandler` or not. But that won't work cause you're right about missing the cleanup part, I guess the `tappedPrimaryButton` flag should be in the superclass or is there a better way to do this?

Reply:
Hmm… I see…

So you may want to change where the dismiss handler is being called in the super class
Or add a flag to the super class to not call the dismiss handler
Or you could clear the button variable so no UI updates happen during dismissal - basically disconnect the UI from the VC
Or you could defer the UI update using a dispatch_async - so the UI update doesn't happen until after the view disappears. 
Also maybe move logic into viewDidDissapear instead?

All or none of those may work… probably have to play around with them to see

Reply:
ended up adding a flag to the superclass, I set it in init because there's at least one case (`wmf_showKeepSavedArticlesOnDevicePanelIfNecessary` in `Panels`, line 228) when we don't want to discard `dismissHandler` simply because we tap on the button (setting it to `true` in `primaryButtonTapped` and then guarding against it in `viewWillDisappear` would prevent `dismissHandler` from happening every time we tap on the button)